doc/vxlan/EVPN/EVPN_VXLAN_HLD.md

```diff
@@ -656,6 +669,9 @@ The following will be added as part of tunnel deletion.
 - sai_tunnel_remove_map, sai_tunnel_remove_tunnel_termination, sai_tunnel_remove_tunnel when the tunnel is to be removed on account of the last entry being removed. 
 - VxlanTunnel object will be deleted.
 
+#### 4.3.2.2 P2MP Tunnel Deletion
```

Comment:
1. The  tunnel object can be deleted only after all the IMR, MAC, IP Prefix routes are removed by BGP. 
    The bridge port object can be deleted only after the fdb count referencing the bridgeport goes down to 0. 

2. Once the map count becomes 0, the vxlanmgr holds off handling further Map creations till the number of P2P 
    tunnels goes down to 0. This is to handle quick delete and readd of mapping entries. 
   VxlanMgr gets to know of P2P tunnel count by looking at the number of state table entries. 
    However for P2MP the state table entries are not populated..    
    How is this scenario handled ?

Reply:
There are no changes to P2MP tunnel logic currently except int this case it will always have 0 dip tunnels. Rest of the logic is the same as in existing code. I did have a question on FDB reference. In current orchagent code i don't not see tunnel delete skipped based on fdb reference count. Only bridge port removal is skipped while the actual tunnel might be deleted even when FDB reference count is non zero. Is that by design? If that's the case isn't it an issue with bridge port still referencing the tunnel port? https://github.com/Azure/sonic-swss/blob/4f1d726d4cbf8a283b22cd5f612cf03ca21a27b3/orchagent/vxlanorch.cpp#L1499

Reply:
Few more questions
1) I don't see TUNNEL_USER_MAC being used anywhere. Is it same as fdb count?
2) I believe IP prefix routes use the P2MP tunnel rather than P2P tunnel but their references are stored in P2P tunnels. In this scenario since there are no P2P tunnels, I believe i should have route references on P2MP tunnel itself. Am I right here?
3) When IMR is received the tunnel's bridge port will be added to L2MC group member. So maintaining a reference here and deleting the bridge port and in turn tunnel only when IMR count reaches zero. This will be identical to DIP tunnel reference used. Is this fine?

Reply:
1. m_fdb_count is not the same as the evpn mac count. However currently the evpn mac ref cnt is not being used. 
2. The refcnts were maintained for P2P tunnels to create and delete the P2P tunnels. The DIP tunnel count was used for deciding when to delete the P2MP tunnel. 
3. For the P2MP only case the refcnts could be reused to decide when to delete the P2MP tunnel object. Please check for del_tnl_hw_pending and getDipTunnelCnt and deletePendingSIPTunnel.  On the last map entry deletion the P2MP tunnel object will be deleted once the total of refcnts across all the remote VTEP IP becomes zero. 
4. For bridgeport deletion.. the Port object and Associated bridgeport object (P2P) is deleted only when the total fdb count goes down to 0. It is after the bridge port deletion do we delete the Tunnel Object as well. 
5. My comment was that these are applicable to P2MP scenarios also and needs to be mentioned in the HLD. Code wise we may have to do some more adjustments to reuse the existing implementation for P2MP as well. 
6. Regarding the L2MC group member yes using the existing IMR refcnt to decide when to add/remove the L2MC group members would be good.

Reply:
Regarding your query https://github.com/Azure/SONiC/pull/806/files#r669059056  ,  No the bridgeport is deleted first and then the tunnel object which is being referenced by the bridgeport. 
As part of deleteDynamicDIPTunnel there is a call to getTunnelPort. If this exists then the Tunnel Object is not deleted.

Reply:
@dgsudharsan @srj102 Currently when tunnel nexthops are created by orchagent, it doesn't check for reachability of remote destination. Are we following the same approach or add the reachability condition while creating the the nexthops?

Reply:
I don't think we are enhancing that here.  The existing approach still holds true.